raphtory/src/algorithms/embeddings/fast_rp.rs

```diff
@@ -0,0 +1,427 @@
+use crate::{
+    algorithms::algorithm_result::AlgorithmResult,
+    core::{entities::VID, state::compute_state::ComputeStateVec},
+    db::{
+        api::view::{NodeViewOps, StaticGraphViewOps},
+        task::{
+            context::Context,
+            node::eval_node::EvalNodeView,
+            task::{ATask, Job, Step},
+            task_runner::TaskRunner,
+        },
+    },
+};
+use rand::prelude::*;
+use rayon::prelude::*;
+use std::sync::Arc;
+
+#[derive(Clone, Debug, Default)]
+struct FastRPState {
+    embedding_state: Vec<f64>,
+}
+
+/// Computes the embeddings of each vertex of a graph using the Fast RP algorithm
+///
+/// # Arguments
+///
+/// * `graph` - A reference to the graph
+/// * `embedding_dim` - The size of the generated embeddings
+/// * `normalization_strength` - The extent to which high-degree vertices should be discounted (range: 1-0)
+/// * `iter_weights` - The scalar weights to apply to the results of each iteration
+/// * `seed` - The seed for initialisation of random vectors
+/// * `threads` - Number of threads to use
+///
+/// Returns:
+///
+/// An AlgorithmResult containing the mapping from the node to its embedding
+///
+pub fn fast_rp<G>(
+    graph: &G,
+    embedding_dim: usize,
+    normalization_strength: f64,
+    iter_weights: Vec<f64>,
+    seed: Option<u64>,
+    threads: Option<usize>,
+) -> AlgorithmResult<G, Vec<f64>, Vec<f64>>
+where
+    G: StaticGraphViewOps,
+{
+    let ctx: Context<G, ComputeStateVec> = graph.into();
+    let m = graph.count_nodes() as f64;
+    let s = m.sqrt();
+    let beta = normalization_strength - 1.0;
+    let num_iters = iter_weights.len() - 1;
+    let weights = Arc::new(iter_weights);
+    let seed = seed.unwrap_or(rand::thread_rng().gen());
+
+    // initialize each vertex with a random vector according to FastRP's construction rules
+    let step1 = {
+        let weights = Arc::clone(&weights);
+        ATask::new(move |vv| {
+            let l = ((vv.degree() as f64) / (m * 2.0)).powf(beta);
+            let choices = [
+                (l * s.sqrt(), 1.0 / (s * 2.0)),
+                (0.0, 1.0 - (1.0 / s)),
+                (-l * s.sqrt(), 1.0 / (s * 2.0)),
+            ];
+            let mut rng = SmallRng::seed_from_u64(vv.node.0 as u64 ^ seed);
+            let state: &mut FastRPState = vv.get_mut();
+            state.embedding_state = (0..embedding_dim)
+                .map(|_| choices.choose_weighted(&mut rng, |item| item.1).unwrap().0 * weights[0])
+                .collect();
+            Step::Continue
+        })
+    };
+
+    // sum each vector from neighbours and scale
+    let step2 = ATask::new(move |vv: &mut EvalNodeView<G, FastRPState>| {
+        // for neighbor, for i, add neighbors.prev[i] to current state
+        // scale state by iteration weight
+        let weights = Arc::clone(&weights);
+        let ss = vv.eval_graph.ss;
+        // TODO: rewrite using iters?
+        for neighbour in vv.neighbours() {
```

Comment:
this algorithm is treating the network as undirected, would a directed version make sense?

Reply:
The base implementation is only defined for undirected/bidirectional graphs, so this is correct as is. There is, however, a modification I intend to make in the future where different segments of the embedding vector encode different edge types and/or features.